include/macis/gf/bandlan.hpp

```diff
@@ -0,0 +1,282 @@
+ *
+ * @returns bool: Error code from LAPACK.
+ *
+ * @author Carlos Mejuto Zaera
+ * @date 25/04/2022
+ */
+bool GetEigsysBand(std::vector<double> &mat, int nSupDiag,
+                   std::vector<double> &eigvals, std::vector<double> &eigvecs,
+                   int matsize);
+
+/**
+ * @brief Perform a band Lanczos calculation on the Hamiltonian operator H,
+ * starting from vectors qs, for at most nLanIts iterations. The resulting
+ * band-diagonal matrix Hamiltonian will be stored in bandH. Note that this
+ * implementation does not account for deflations (i.e., pruning the span of the
+ * qs for linear dependencies in higher powers of H).
+ *
+ * @param[in] const sparseexx::csr_matrix<double, int32_t> &H: Hamiltonian
+ * oprator. Just needs to implement a matrix vector product.
+ * @param[in] std::vector<Cont> &qs: Initial set of vetors to
+ * perform the band Lanczos on. Deleted on exit.
+ * @param[in] std::vector<Cont> &bandH: On exit, band-diagonal
+ * Hamiltonian approximation.
+ * @param[inout] int &nLanIts: Number of Lanczos iterations to perform.
+ * @param[in] double thres: Threshold determining when to ignore beta's for
+ * being too small.
+ * @param[in] bool print: If true, write intermediate results to file.
+ *
+ * @author Carlos Mejuto Zaera
+ * @date 25/04/2022
+ */
+template <class Cont, class Functor>
+void MyBandLan(const Functor &H, std::vector<Cont> &qs,
+               std::vector<Cont> &bandH, int &nLanIts, int nbands, int N,
+               double thres = 1.E-6, bool print = false) {
+  // BAND LANCZOS ROUTINE. TAKES AS INPUT THE HAMILTONIAN H, INITIAL VECTORS qs
+  // AND RETURNS THE BAND HAMILTONIAN bandH. IT PERFORMS nLanIts ITERATIONS,
+  // STOPPING IF THE NORM OF ANY NEW KRYLOV VECTOR IS BELOW thres. IF LANCZOS IS
+  // STOPPED PREMATURELY , nLanIts IS OVERWRITTEN WITH THE ACTUAL NUMBER OF
+  // ITERATIONS! THE qs VECTOR IS ERASED AT THE END OF THE CALCULATION
+  using dbl = std::numeric_limits<double>;
+  bandH.clear();
+  bandH.resize(nLanIts * nLanIts, 0.);
+
+  // MAKE SPACE FOR 2 * nbands VECTORS
+  qs.resize(2 * nbands * N);
+  std::vector<Cont> temp(N, 0.);
+  if(print) {
+    for(int i = 0; i < nbands; i++) {
+      std::ofstream ofile("lanvec_" + std::to_string(i + 1) + ".dat",
+                          std::ios::out);
+      ofile.precision(dbl::max_digits10);
+      for(size_t el = 0; el < N; el++)
+        ofile << std::scientific << qs[el + i * N] << std::endl;
+      ofile.close();
+    }
+  }
+  // DICTIONARY TO KEEP THE REAL INDICES OF THE KRYLOV VECTORS
+  // THIS IS NECESSARY IN ORDER TO ONLY STORE 2* nbands OF THEM
+  // AT ANY POINT IN TIME, PLUS ONE SCRATCH VECTOR TO BE DEFINED
+  // INSIDE THE FOR LOOP
+  std::vector<int> true_indx(nLanIts + 1);
+  for(int i = 0; i < nbands; i++) true_indx[i + 1] = i;
+  int next_indx = nbands;
+
+  for(int it = 1; it <= nLanIts; it++) {
+    int band_indx_i =
+        true_indx[it];  // TO WHAT ELEMENT OF THE VECTOR SET DO WE APPLY THIS
```

Comment:
Can you explain this indirection here? Generally we want to avoid things like this (to e.g. allow for better usage of Level-3 BLAS when possible), but if its completely necessary, it's fine.

Reply:
You mean the use of the effective index `true_indx`? (I'm afraid I don't know what you mean by indirection)

This is what I came up with to apply the matvecs in the right order to the right vector, since unlike in regular Lanczos at every iteration you act on a different vector in your set of `nbands`. Why is this a problem for BLAS? Shouldn't it be fine as long as you pass the right memory sector to act on?

I guess that a way to avoid this `true_indx` variable could be to turn the for loop over `it` into two, one external loop going over `nLanIts // nbands` and one internal loop going over `nbands`. I think that should be completely equivalent, and then the index in the internal loop would be essentially iterating over which vector in the basis to act on with the Hamiltonian. Do you think this would make the code better?